### DISCUSSION TITLE
Components Example

### DISCUSSION THREAD

**Discussion by Simonl9l** [nuke-build/nuke#1394] [Q&A]
Trying to understand components. Look like it has lots of potential for our use case (s) - lot's of similar builds!

per the [Parameters](https://nuke.build/docs/sharing/build-components/#parameters) I'm looking to access the solution, but `  [Solution] Solution Solution => TryGetValue(() => Solution)!;` comes back with null?

I see in the link above, that `Solution` is defined as `string`, however that mean I can then get to the `Solution` properties?

Am I also supposed to define `[Solution] Solution Solution;` in the main `Build.cs` file too?

I'm getting a `No solution file defined for 'Solution'.` exception?

here is the Component:
```C#
[ParameterPrefix(nameof(IBuildRestore))]
public interface IBuildRestore : INukeBuild
{
    [Solution] Solution Solution => TryGetValue(() => Solution)!;
 
    public Target Restore => _ => _
        .Executes(() =>
    {
        Log.Debug(Solution);
    }); 
```
here is the build:
```C#
class Build : NukeBuild, IBuildRestore 
{
    public static int Main () => Execute<Build>(x => x.DoBuild);
   
    [Solution] Solution Solution;
        
    public Target DoBuild => _ => _
        .DependsOn<IBuildRestore>(t => t.Restore)
        .Executes(() =>
        {
            Log.Debug(Solution);
        });
}
```
here is the error:
```
12:45:37 [WRN] Could not inject value for IBuildRestore.Solution
System.ArgumentException: No solution file defined for 'Solution'.
   at Nuke.Common.Assert.NotNull[T](T obj, String message, String argumentExpression) in /_/source/Nuke.Utilities/Assert.cs:line 73
   at Nuke.Common.ProjectModel.SolutionAttribute.GetSolutionFileFromParametersFile(MemberInfo member) in /_/source/Nuke.Common/Attributes/SolutionAttribute.cs:line 83
   at Nuke.Common.ProjectModel.SolutionAttribute.GetValue(MemberInfo member, Object instance) in /_/source/Nuke.Common/Attributes/SolutionAttribute.cs:line 53
   at Nuke.Common.ValueInjection.ValueInjectionAttributeBase.TryGetValue(MemberInfo member, Object instance) in /_/source/Nuke.Build/Execution/Extensibility/ValueInjectionAttributeBase.cs:line 26

```

from the Build - the `Solution` is correctly returned:
```
[DBG] /{file system path}/NukeConstructs.sln

```

Is there a comprehensive working example I can crib from - as Im obviously missing something from the documentation?

On review of [this](https://github.com/nuke-build/nuke/blob/85998da9b9cc713af50f5ffbc9582323be812ac0/build/Build.cs#L62C4-L62C73), it seems that this is directly setting the interfaces Solution property, where as the understandings is that is what the `TryGetValue(() => Solution)` is supposed to do?

**Comment by StefanOssendorf** [comment]
Hi. Did you solve this problem?

**Comment by Simonl9l** [comment]
@StefanOssendorf I did, not sire how at this pint specifically, but copied examples I found in other repos in GitHub.

**Reply by StefanOssendorf** [reply]
Hi. Is it possible for you to share your solution/implementation here or a link to your sources (if available)?
